spark/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala

```diff
@@ -439,15 +439,28 @@ class DeltaLog private(
   def ensureLogDirectoryExist(): Unit = {
     val fs = logPath.getFileSystem(newDeltaHadoopConf())
     def createDirIfNotExists(path: Path): Unit = {
-      var success = false
-      try {
-        success = fs.exists(path) || fs.mkdirs(path)
+      // Optimistically attempt to create the directory first without checking its existence.
+      // This is efficient because we're assuming it's more likely that the directory doesn't
+      // exist and it saves an filesystem existence check in that case.
+      val (success, mkdirsIOExceptionOpt) = try {
+        // Return value of false should mean the directory already existed (not an error) but
+        // we will verify below because we're paranoid about buggy FileSystem implementations.
+        (fs.mkdirs(path), None)
       } catch {
+        // Only needed because buggy Hadoop FileSystem.mkdir wrongly throws on existing dir.
```

Comment:
This is tricky... if a not-directory object already exists with the same name, then a `FileAlreadyExistsException` can and should be thrown. The bug is that HDFS _also_ throws if the already-existing object is a proper directory (so we have to double check whether the exception was wrongly thrown).

Reply:
Updated the comment.